src/Microsoft.OData.Core/Json/ODataUtf8JsonWriter.cs

```diff
@@ -377,6 +384,99 @@ public void WriteValue(string value)
             this.FlushIfBufferThresholdReached();
         }
 
+        private void WriteStringValueInChunks(ReadOnlySpan<char> value)
+        {
+            int chunkSize = bufferWriter.FreeCapacity;
+
+            for (int i = 0; i < value.Length; i += chunkSize)
+            {
+                int remainingChars = Math.Min(chunkSize, value.Length - i);
+                ReadOnlySpan<char> chunk = value.Slice(i, remainingChars);
+
+                if (NeedsEscaping(chunk) != -1)
+                {
+                    WriteEscapedStringChunk(chunk);
+                }
+                else
+                {
+                    WriteStringChunk(chunk);
+                }
+
+                // Flush the buffer if needed
+                this.FlushIfBufferThresholdReached();
+            }
+        }
+
+        private int NeedsEscaping(ReadOnlySpan<char> value)
+        {
+            if (value.IsEmpty)
+            {
+                return -1;
+            }
+
+            // Iterate through each character in the span
+            for (int i = 0; i < value.Length; i++)
+            {
+                string encoded = this.encoder.Encode(value[i].ToString());
+                
+                // if encoded string is different from original string then we know special
+                // characters exist and need escaping.
+                if (!string.Equals(value[i].ToString(), encoded, StringComparison.Ordinal))
+                {
+                    return i;
+                }
+            }
+
+            return -1;
+        }
+
+        private void WriteEscapedStringChunk(ReadOnlySpan<char> sourceChunk)
+        {
+            char[] valueArray = null;
+            const int maxSize = 0; 
+            const int threshold = 120;
+
+            Span<char> destination = maxSize <= threshold
+                ? stackalloc char[maxSize]
+                :valueArray = ArrayPool<char>.Shared.Rent(maxSize);
+
+            try
+            {
+                OperationStatus status = this.encoder.EncodeUtf8(
+                    MemoryMarshal.Cast<char, byte>(sourceChunk),
+                    MemoryMarshal.Cast<char, byte>(destination), // Use the byte span of the destination
+                    out int bytesConsumed,
+                    out int bytesWritten,
+                    isFinalBlock: true
+                );

```

Comment:
[EncodeUtf8](https://learn.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.memorymarshal.cast?view=net-8.0) expects the input to already be utf-8 text. Casting char span to byte span does not seem safe. I'm almost certain if we interpret the utf-16 text as utf-8 data without transcoding, we may end up with corrupted data.

We should probably be using the [`JavaScriptEncoder.Encode`](https://learn.microsoft.com/en-us/dotnet/api/system.text.encodings.web.textencoder.encode?view=net-7.0#system-text-encodings-web-textencoder-encode(system-readonlyspan((system-char))-system-span((system-char))-system-int32@-system-int32@-system-boolean)) method instead which takes a `ReadOnlySpan<char>` input:

```c#
OperatioStatus status = this.encoder.Encode(
  chunk,
 destination,
 out int charsConsumed,
 out int bytesWritten,
 isFinalBlock: false /* this should be set to true on the last chunk of the string, so should be passed as parameter to this method */
);

Debug.Assert(status == OperationStatus.Done);
Debug.Assert(charsConsumed == chunk.Length);
WriteStringChunk(destination.Slice(0, bytesWritten));

// ...
```